### PR TITLE
Prisma 2: TypeScript/JavaScript Client API

### PR DESCRIPTION
**Pull request by schickling**
[Rendered view](https://github.com/prisma/rfcs/blob/new-ts-client-rfc/text/0000-new-ts-client.md)

### CODE REVIEW

**Comment by ejoebstl** [comment]
Hello, please give a short roundup about the new `select`. 
* Why is it needed? 
* How does it map to the GraphQL wire protocol?

**Comment by schickling** [comment]
The new `select` field was introduced to avoid the following inconsistencies having to do with 2 positional arguments and some of the resulting edge cases:

- Remove the need for `$` prefixing (e.g. `$first`) in the selection set
- Remove the need for empty objects when no modifying arguments (`where`, `first` etc) are needed. Example: `prisma.users({}, { posts: true })`

The goal was to reach consistency in how you're constructing a query (regardless if top-level or nested).

**Comment by ejoebstl** [comment]
Hello, 

can you please add an example of a nested filter (`where`) using the new `select` interface?

**Comment by schickling** [comment]
Yes, I've added an example in my latest commit. Here is a copy:

```ts
const dynamicResult1x: DynamicResult1 = await prisma.users.findOne({
    where: 'bobs-id',
    select: {
      posts: {
        where: { title_contains: 'Prisma' },
        select: { comments: true }
      },
      friends: true
    }
  })
```

**Comment by joshhopkins** [comment]
Would be great if this was considered for v2: https://github.com/prisma/prisma/issues/4075

**Comment by matthewmueller** [comment]
```ts
 // Ordering
  const usersByEmail = await prisma.users({ orderBy: { email: 'ASC' } })
  const usersByEmailAndName = await prisma.users({
    orderBy: [{ email: 'ASC' }, { name: 'DESC' }],
  })
  const usersByProfile = await prisma.users({
    orderBy: { profile: { imageSize: 'ASC' } },
  })
```

How does this work with selects? Example in SQL:

```
select * from users where name = 'Johannes' order by age desc;
```

**Comment by schickling** [comment]
Why wouldn't this work?
```ts
const usersByProfile = await prisma.users({
  orderBy: { profile: { imageSize: 'ASC' } },
  where: { name: 'Johannes' },
})
```

**Comment by matthewmueller** [comment]
Oh it does, just wasn't specified in the spec, so I wasn't sure if it was the way you mentioned or if it was:

```js
const usersByProfile = await prisma.users({
  orderBy: { profile: { imageSize: 'ASC' } },
  name: 'Johannes'
})
```

Needing this example for the Go spec 😄

**Comment by matthewmueller** [comment]
One potential issue with this API that I ran into with the Go Client is the following:

```ts
const dynamicResult3: DynamicResult3 = await prisma.users.findOne({
    where: 'bobs-id',
    select: { posts: { select: { aggregate: { count: true } } } },
  })
```

`aggregate` could potentially be the name of a model. My thought is that this is probably fine, but then the case could also be made that `withPageInfo`, `stream`, etc. could be considered reserved.

**Comment by schickling** [comment]
Generally this is certainly still weird and we should reconcile the inconsistency of using `$` vs not using it. The downside of not using the `$` "marker" is that it could be confusing whether something is a data field vs a special feature.

Re `aggregate`: Right now the idea is to throw an error during client generation time and allowing the user to overwrite the conflicting field name via the generator modifier.

**Comment by matthewmueller** [comment]
One possible solution for `$stream` is by entering "different modes" upfront, we can enter a "streaming" mode or an "iterator" mode. These will affect what the "thenables" return.

Here's a psuedo-code example:

```ts

import Stream from 'prisma/stream'

// Streaming mode
const stream: Stream<Post> = await Stream(opts)
  .users
  .findOne('bobs-id')
  .posts({ first: 50 })

// Streaming data
// TODO: not sure of exact API
for await (const post of stream) {
  console.log(post)
}

// iterator

import Iterator from 'prisma/iterator'

// Iterator mode
const iter: Iterator<Post> = await Iterator(opts)
  .users
  .findOne('bobs-id')
  .posts({ first: 50 })

// Iterate over data
// TODO: not sure of exact API
while (!iter.done) {
  const post = (await iter.next()).value
}
```

**Comment by matthewmueller** [comment]
Not sure I understand the `name` key in this API:

```ts
// Lookup-by Multi-field indexes
const john: User = await prisma.users.findOne({
  name: { firstName: 'John', lastName: 'Doe' },
})
```

Is that the name of `unique(first_name, last_name)` constraint?

What do you guys think about detecting these cases and adding methods like:

```ts
await prisma.users.findByEmail("alice@bob.com")
await prisma.users.findByFirstNameAndLastName("John", "Doe")
```

?

**Comment by matthewmueller** [comment]
One interesting conclusion that @schickling and I came to today was the need to differentiate between `undefined`, `null` and `some value`.

Given a datamodel like this:

```groovy
model User {
  id: ID!
  email: String!
  first_name: String
}
```

`first_name` can be either `null` or `text` in the database. Supported operations in SQL include:

```sql
-- inserts
insert into users (email, first_name) values ('alice@prisma.io', 'Alice');
insert into users (email, first_name) values ('alice@prisma.io', null);

-- selects
select * users from users where first_name = 'Alice';
select * users from users where first_name is null;
```

From the client's perspective, we need to be careful to differentiate between `undefined`, `null`, and `some value`. 

Currently we don't really do this. In Typescript, we need to be more restrictive in our type definitions for required properties:

```diff
type User = {
- email?: String | null
+ email?: String
}
``` 

Where this query no longer works:

```
await prisma.users({ where: { email: null } })
```

You'd have to do:

```ts
await prisma.users({ where: { email: undefined } })
// or
await prisma.users({ where: {} })
```

And where the following queries would do different things:

```ts
await prisma.users({ where: { first_name: undefined } })
await prisma.users({ where: { first_name: null } })
await prisma.users({ where: { first_name: "Alice" } })
```

**Comment by P4sca1** [comment]
Would be great if the prisma client would support caching / batching to solve the n+1 query problem.
While the prisma graphql server internally already uses `dataloader` to achieve this, `prisma-client` has no way to solve this problem atm.
I outlined some of my thoughts [here](https://github.com/prisma/prisma/issues/3599).

I will sum up my idea below.

**Background:**
I am using `prisma` in my resolvers to implement my own graphql server using `apollo-server`.

**Problem:**
Consider the following query taken from the `dataloader` example.

```gql
{
  me {
    name
    bestFriend {
      name
    }
    friends(first: 5) {
      name
      bestFriend {
        name
      }
    }
  }
}
```

This query will lead to a maximum of 8 calls to `prisma`.  1 for `me`, 1 for `me.bestFriend`, 1 for `me.friends` and 5 for `me.friends.bestFriend`.
`dataloader` reduces this amount by
a) automatically batching all the calls to `bestFriend` into one query.
b) caching `bestFriend`s that are already requested.

There is already an open issue to support automatic batching in the client implementation [here](https://github.com/prisma/prisma/issues/3599), so what's left is caching. The reason we need caching is that we are not able to batch all of our prisma requests at once, because resolvers are not executed at the same time and there may be multiple sequentially executed prisma requests in one resolver.

Caching would work on a per graphql request basis. This is what I have in mind:
```ts
new ApolloServer({
  context: () => {
    // Prisma client with its own in memory cache valid only during the lifetime of the request.
    // Will be garbage-collected after the request was processed.
    prisma: prisma.dataloader()
  }
}
```

Basically we need to wrap the `prisma` client into a `dataloader`, which would allow for automatic batching and caching. The biggest problem I see here is that fetching in `dataloader` works by knowing the ids to request before doing the actual request. What this essentially means for the above query is that we need something like `me.bestFriendId` and `me.friendIds` after fetching `me` to allow fetching the rest of the relations using `dataloader`.

Hope to get some input on this and how you guys solve the n+1 query problem when implementing your own graphql server using `prisma-client`.

**Comment by schickling** [comment]
Closing in favor of https://github.com/prisma/specs/pull/5
